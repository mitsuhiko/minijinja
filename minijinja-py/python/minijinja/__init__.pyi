from pathlib import PurePath
from typing import (
    Any,
    Callable,
    Literal,
    TypeVar,
    Protocol,
    overload,
)
from typing_extensions import Final, TypeAlias, Self
from minijinja._lowlevel import State
from collections.abc import Mapping

__all__ = [
    "Environment",
    "TemplateError",
    "safe",
    "escape",
    "render_str",
    "eval_expr",
    "pass_state",
]

_A_contra = TypeVar("_A_contra", contravariant=True)
_R_co = TypeVar("_R_co", covariant=True)

class _PassesState(Protocol[_A_contra, _R_co]):
    def __call__(self, state: State, value: _A_contra, /) -> _R_co: ...

    __minijinja_pass_state__: Literal[True]

_StrPath: TypeAlias = PurePath | str
_Behavior = Literal["strict", "lenient", "chainable"]

DEFAULT_ENVIRONMENT: Final[Environment]

def render_str(source: str, name: str | None = None, /, **context: Any) -> str: ...
def eval_expr(expression: str, /, **context: Any) -> Any: ...

class Environment:
    loader: Callable[[str], str] | None
    fuel: int | None
    debug: bool
    undefined_behavior: _Behavior
    auto_escape_callback: Callable[[str], bool] | None
    path_join_callback: Callable[[str, str], _StrPath] | None
    keep_trailing_newline: bool
    trim_blocks: bool
    lstrip_blocks: bool
    finalizer: _PassesState[Any, Any] | Callable[[Any], Any] | None
    reload_before_render: bool
    block_start_string: str
    block_end_string: str
    variable_start_string: str
    variable_end_string: str
    comment_start_string: str
    comment_end_string: str
    line_statement_prefix: str | None
    line_comment_prefix: str | None

    @overload
    def __init__(
        self,
        loader: Callable[[str], str] | None = None,
        templates: Mapping[str, str] | None = None,
        filters: Mapping[str, Callable[[Any], Any]] | None = None,
        tests: Mapping[str, Callable[[Any], bool]] | None = None,
        globals: Mapping[str, Any] | None = None,
        debug: bool = True,
        fuel: int | None = None,
        undefined_behavior: _Behavior = "lenient",
        auto_escape_callback: Callable[[str], bool] | None = None,
        path_join_callback: Callable[[str, str], _StrPath] | None = None,
        keep_trailing_newline: bool = False,
        trim_blocks: bool = False,
        lstrip_blocks: bool = False,
        finalizer: _PassesState[Any, Any] | None = None,
        reload_before_render: bool = False,
        block_start_string: str = "{%",
        block_end_string: str = "%}",
        variable_start_string: str = "{{",
        variable_end_string: str = "}}",
        comment_start_string: str = "{#",
        comment_end_string: str = "#}",
        line_statement_prefix: str | None = None,
        line_comment_prefix: str | None = None,
    ) -> None: ...
    @overload
    def __init__(
        self,
        loader: Callable[[str], str] | None = None,
        templates: Mapping[str, str] | None = None,
        filters: Mapping[str, Callable[[Any], Any]] | None = None,
        tests: Mapping[str, Callable[[Any], bool]] | None = None,
        globals: Mapping[str, Any] | None = None,
        debug: bool = True,
        fuel: int | None = None,
        undefined_behavior: _Behavior = "lenient",
        auto_escape_callback: Callable[[str], bool] | None = None,
        path_join_callback: Callable[[str, str], _StrPath] | None = None,
        keep_trailing_newline: bool = False,
        trim_blocks: bool = False,
        lstrip_blocks: bool = False,
        finalizer: Callable[[Any], Any] | None = None,
        reload_before_render: bool = False,
        block_start_string: str = "{%",
        block_end_string: str = "%}",
        variable_start_string: str = "{{",
        variable_end_string: str = "}}",
        comment_start_string: str = "{#",
        comment_end_string: str = "#}",
        line_statement_prefix: str | None = None,
        line_comment_prefix: str | None = None,
    ) -> None: ...
    def remove_filter(self, name: str) -> None: ...
    def add_test(self, name: str, test: Callable[[Any], bool]) -> None: ...
    def remove_test(self, name: str) -> None: ...
    def add_global(self, name: str, value: Any) -> None: ...
    def remove_global(self, name: str) -> None: ...
    def render_template(self, template_name: str, /, **context: Any) -> str: ...
    def render_str(
        self, source: str, name: str | None = None, /, **context: Any
    ) -> str: ...
    def eval_expr(self, expression: str, /, **context: Any) -> Any: ...

class TemplateError(RuntimeError):
    def __init__(self, message: str) -> None: ...
    @property
    def message(self) -> str: ...
    @property
    def kind(self) -> str: ...
    @property
    def name(self) -> str | None: ...
    @property
    def detail(self) -> str | None: ...
    @property
    def line(self) -> int | None: ...
    @property
    def range(self) -> int | None: ...
    @property
    def template_source(self) -> str | None: ...
    def __str__(self) -> str: ...

class Markup(str):
    def __html__(self) -> Self: ...

def safe(value: str) -> str: ...
def escape(value: Any) -> str: ...
def pass_state(
    f: Callable[[State, _A_contra], _R_co],
) -> _PassesState[_A_contra, _R_co]: ...
