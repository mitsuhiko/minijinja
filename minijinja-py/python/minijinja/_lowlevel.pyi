from typing import Any
from . import Environment
from typing_extensions import final

@final
class State:
    name: str
    env: Environment
    current_block: str | None
    auto_escape: bool

    def lookup(self, name: str) -> Any: ...
