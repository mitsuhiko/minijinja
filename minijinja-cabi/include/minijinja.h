/*
 * MiniJinja for C
 *
 * Copyright 2024 Armin Ronacher. MIT License.
 */

#ifndef _minijinja_h_included
#define _minijinja_h_included

#pragma once

#include <stdint.h>
#include <stddef.h>
#include <stdbool.h>

#ifndef MINIJINJA_API
#  define MINIJINJA_API
#endif


/*
 The kind of error that occurred.
 */
typedef enum mj_err_kind {
  MJ_ERR_KIND_NON_PRIMITIVE,
  MJ_ERR_KIND_NON_KEY,
  MJ_ERR_KIND_INVALID_OPERATION,
  MJ_ERR_KIND_SYNTAX_ERROR,
  MJ_ERR_KIND_TEMPLATE_NOT_FOUND,
  MJ_ERR_KIND_TOO_MANY_ARGUMENTS,
  MJ_ERR_KIND_MISSING_ARGUMENT,
  MJ_ERR_KIND_UNKNOWN_FILTER,
  MJ_ERR_KIND_UNKNOWN_FUNCTION,
  MJ_ERR_KIND_UNKNOWN_TEST,
  MJ_ERR_KIND_UNKNOWN_METHOD,
  MJ_ERR_KIND_BAD_ESCAPE,
  MJ_ERR_KIND_UNDEFINED_ERROR,
  MJ_ERROR_KIND_BAD_SERIALIZTION,
  MJ_ERR_KIND_BAD_INCLUDE,
  MJ_ERR_KIND_EVAL_BLOCK,
  MJ_ERR_KIND_CANNOT_UNPACK,
  MJ_ERR_KIND_WRITE_FAILURE,
  MJ_ERR_KIND_UNKNOWN,
} mj_err_kind;

/*
 Controls the undefined behavior of the engine.
 */
typedef enum mj_undefined_behavior {
  /*
   The default, somewhat lenient undefined behavior.
   */
  MJ_UNDEFINED_BEHAVIOR_LENIENT,
  /*
   Complains very quickly about undefined values.
   */
  MJ_UNDEFINED_BEHAVIOR_STRICT,
  /*
   Like Lenient, but also allows chaining of undefined lookups.
   */
  MJ_UNDEFINED_BEHAVIOR_CHAINABLE,
} mj_undefined_behavior;

/*
 The kind of a value.
 */
typedef enum mj_value_kind {
  MJ_VALUE_KIND_UNDEFINED,
  MJ_VALUE_KIND_NONE,
  MJ_VALUE_KIND_BOOL,
  MJ_VALUE_KIND_NUMBER,
  MJ_VALUE_KIND_STRING,
  MJ_VALUE_KIND_BYTES,
  MJ_VALUE_KIND_SEQ,
  MJ_VALUE_KIND_MAP,
  MJ_VALUE_KIND_ITERABLE,
  MJ_VALUE_KIND_PLAIN,
  MJ_VALUE_KIND_INVALID,
} mj_value_kind;

/*
 Pointer to a MiniJinja environment.
 */
typedef struct mj_env mj_env;

/*
 Helps iterating over a value.
 */
typedef struct mj_value_iter mj_value_iter;

/*
 Opaque value type.
 */
typedef struct mj_value {
  uintptr_t _opaque[3];
} mj_value;

/*
 Allows one to override the syntax elements.
 */
typedef struct mj_syntax_config {
  const char *block_start;
  const char *block_end;
  const char *variable_start;
  const char *variable_end;
  const char *comment_start;
  const char *comment_end;
  const char *line_statement_prefix;
  const char *line_comment_prefix;
} mj_syntax_config;

#ifdef __cplusplus
extern "C" {
#endif // __cplusplus

/*
 Registers a template with the environment.
 */
MINIJINJA_API bool mj_env_add_template(struct mj_env *env, const char *name, const char *source);

/*
 Clears all templates.
 */
MINIJINJA_API bool mj_env_clear_templates(struct mj_env *env);

/*
 Evaluate an expression.
 */
MINIJINJA_API
struct mj_value mj_env_eval_expr(const struct mj_env *env,
                                 const char *expr,
                                 struct mj_value ctx);

/*
 Frees a MiniJinja environment.
 */
MINIJINJA_API void mj_env_free(struct mj_env *env);

/*
 Allocates a new and empty MiniJinja environment.
 */
MINIJINJA_API struct mj_env *mj_env_new(void);

/*
 Removes a template from the environment.
 */
MINIJINJA_API bool mj_env_remove_template(struct mj_env *env, const char *name);

/*
 Renders a template from a named string.
 */
MINIJINJA_API
char *mj_env_render_named_str(const struct mj_env *env,
                              const char *name,
                              const char *source,
                              struct mj_value ctx);

/*
 Renders a template registered on the environment.
 */
MINIJINJA_API
char *mj_env_render_template(const struct mj_env *env,
                             const char *name,
                             struct mj_value ctx);

/*
 Enables or disables debug mode.
 */
MINIJINJA_API void mj_env_set_debug(struct mj_env *env, bool val);

/*
 Preserve the trailing newline when rendering templates.
 */
MINIJINJA_API void mj_env_set_keep_trailing_newline(struct mj_env *env, bool val);

/*
 Enables or disables the `lstrip_blocks` feature.
 */
MINIJINJA_API void mj_env_set_lstrip_blocks(struct mj_env *env, bool val);

/*
 Changes the recursion limit.
 */
MINIJINJA_API void mj_env_set_recursion_limit(struct mj_env *env, uint32_t val);

/*
 Reconfigures the syntax.
 */
MINIJINJA_API
bool mj_env_set_syntax_config(struct mj_env *env,
                              const struct mj_syntax_config *syntax);

/*
 Enables or disables the `trim_blocks` feature.
 */
MINIJINJA_API void mj_env_set_trim_blocks(struct mj_env *env, bool val);

/*
 Reconfigures the undefined behavior.
 */
MINIJINJA_API
void mj_env_set_undefined_behavior(struct mj_env *env,
                                   enum mj_undefined_behavior val);

/*
 Clears the current error.
 */
MINIJINJA_API void mj_err_clear(void);

/*
 Returns the error's description if there is an error.
 */
MINIJINJA_API const char *mj_err_get_detail(void);

/*
 Returns the error's kind
 */
MINIJINJA_API enum mj_err_kind mj_err_get_kind(void);

/*
 Returns the error's current line.
 */
MINIJINJA_API uint32_t mj_err_get_line(void);

/*
 Returns the error's current template.
 */
MINIJINJA_API const char *mj_err_get_template_name(void);

/*
 Returns `true` if there is currently an error.
 */
MINIJINJA_API bool mj_err_is_set(void);

/*
 Prints the error to stderr.
 */
MINIJINJA_API bool mj_err_print(void);

/*
 Frees an engine allocated string.
 */
MINIJINJA_API void mj_str_free(char *s);

/*
 Sets the syntax to defaults.
 */
MINIJINJA_API void mj_syntax_config_default(struct mj_syntax_config *syntax);

/*
 Appends a value to a list
 */
MINIJINJA_API bool mj_value_append(struct mj_value *slf, struct mj_value value);

/*
 Extracts a float from the value
 */
MINIJINJA_API double mj_value_as_f64(struct mj_value value);

/*
 Extracts an integer from the value
 */
MINIJINJA_API int64_t mj_value_as_i64(struct mj_value value);

/*
 Extracts an unsigned integer from the value
 */
MINIJINJA_API uint64_t mj_value_as_u64(struct mj_value value);

/*
 Debug prints a value to stderr
 */
MINIJINJA_API void mj_value_dbg(struct mj_value value);

/*
 Decrements the refcount
 */
MINIJINJA_API void mj_value_decref(struct mj_value *value);

/*
 Looks up an element by an integer index in a list of object
 */
MINIJINJA_API struct mj_value mj_value_get_by_index(struct mj_value value, uint64_t idx);

/*
 Looks up an element by a string index in an object.
 */
MINIJINJA_API struct mj_value mj_value_get_by_str(struct mj_value value, const char *key);

/*
 Looks up an element by a vaue
 */
MINIJINJA_API struct mj_value mj_value_get_by_value(struct mj_value value, struct mj_value key);

/*
 Returns the value kind.
 */
MINIJINJA_API enum mj_value_kind mj_value_get_kind(struct mj_value value);

/*
 Increments the refcount
 */
MINIJINJA_API void mj_value_incref(struct mj_value *value);

/*
 Checks if the value is numeric
 */
MINIJINJA_API bool mj_value_is_number(struct mj_value value);

/*
 Checks if the value is truthy
 */
MINIJINJA_API bool mj_value_is_true(struct mj_value value);

/*
 Ends the iteration and deallocates the iterator
 */
MINIJINJA_API void mj_value_iter_free(struct mj_value_iter *iter);

/*
 Yields the next value from the iterator.
 */
MINIJINJA_API bool mj_value_iter_next(struct mj_value_iter *iter, struct mj_value *val_out);

/*
 Returns the length of the object
 */
MINIJINJA_API uint64_t mj_value_len(struct mj_value value);

/*
 Creates a new boolean value
 */
MINIJINJA_API struct mj_value mj_value_new_bool(bool value);

/*
 Creates a new f32 value
 */
MINIJINJA_API struct mj_value mj_value_new_f32(float value);

/*
 Creates a new f64 value
 */
MINIJINJA_API struct mj_value mj_value_new_f64(double value);

/*
 Creates a new i32 value
 */
MINIJINJA_API struct mj_value mj_value_new_i32(int32_t value);

/*
 Creates a new i64 value
 */
MINIJINJA_API struct mj_value mj_value_new_i64(int64_t value);

/*
 Creates an empty list value
 */
MINIJINJA_API struct mj_value mj_value_new_list(void);

/*
 Creates a new none value.
 */
MINIJINJA_API struct mj_value mj_value_new_none(void);

/*
 Creates an empty object value
 */
MINIJINJA_API struct mj_value mj_value_new_object(void);

/*
 Creates a new string value
 */
MINIJINJA_API struct mj_value mj_value_new_string(const char *s);

/*
 Creates a new u32 value
 */
MINIJINJA_API struct mj_value mj_value_new_u32(uint32_t value);

/*
 Creates a new u64 value
 */
MINIJINJA_API struct mj_value mj_value_new_u64(uint64_t value);

/*
 Creates a new undefined value.
 */
MINIJINJA_API struct mj_value mj_value_new_undefined(void);

/*
 Inserts a key into an object value.
 */
MINIJINJA_API
bool mj_value_set_key(struct mj_value *slf,
                      struct mj_value key,
                      struct mj_value value);

/*
 Inserts a string key into an object value.
 */
MINIJINJA_API
bool mj_value_set_string_key(struct mj_value *slf,
                             const char *key,
                             struct mj_value value);

/*
 Converts the value into a string.
 */
MINIJINJA_API char *mj_value_to_str(struct mj_value value);

/*
 Iterates over the value.
 */
MINIJINJA_API struct mj_value_iter *mj_value_try_iter(struct mj_value value);

#ifdef __cplusplus
}  // extern "C"
#endif  // __cplusplus

#endif  /* _minijinja_h_included */
