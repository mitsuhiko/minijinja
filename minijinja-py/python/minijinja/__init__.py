from . import _lowlevel


__all__ = ["Environment", "safe", "escape", "render_str", "eval_expr"]


class Environment(_lowlevel.Environment):
    """Represents a MiniJinja environment"""
    def __new__(cls, *args, **kwargs):
        # `_lowlevel.Environment` does not accept any arguments
        return super().__new__(cls)

    def __init__(
        self,
        loader=None,
        templates=None,
        filters=None,
        tests=None,
        globals=None,
        debug=True,
        fuel=None,
        auto_escape_callback=None,
    ):
        super()
        if loader is not None:
            if templates:
                raise TypeError("Cannot set loader and templates at the same time")
            self.loader = loader
        elif templates is not None:
            self.loader = dict(templates).get
        if fuel is not None:
            self.fuel = fuel
        if filters:
            for name, callback in filters.items():
                self.add_filter(name, callback)
        if tests:
            for name, callback in tests.items():
                self.add_test(name, callback)
        if globals is not None:
            for name, value in globals.items():
                self.add_global(name, value)
        self.debug = debug
        if auto_escape_callback is not None:
            self.auto_escape_callback = auto_escape_callback


def render_str(__source, __name=None, **context):
    """Shortcut to render a string with the default environment."""
    return Environment().render_str(__source, __name, **context)


def eval_expr(__expr, **context):
    """Evaluate an expression with the default environment."""
    return Environment().eval_expr(__expr, **context)


try:
    from markupsafe import escape, Markup
except ImportError:
    from html import escape as _escape

    class Markup(str):
        def __html__(self):
            return self

    def escape(value):
        callback = getattr(value, "__html__", None)
        if callback is not None:
            return callback()
        return Markup(_escape(str(value)))


def safe(s):
    """Marks a string as safe."""
    return Markup(s)
