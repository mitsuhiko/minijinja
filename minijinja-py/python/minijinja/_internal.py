# This file contains functions that the rust module imports.
from . import TemplateError, safe


def make_error(info):
    # Internal utility function used by the rust binding to create a template error
    # with info object.  We cannot directly create an error on the Rust side because
    # we want to subclass the runtime error, but on the limited abi it's not possible
    # to create subclasses (yet!)
    err = TemplateError(info.description)
    err._info = info
    return err


# used by the rust runtime to mark something as safe
mark_safe = safe
