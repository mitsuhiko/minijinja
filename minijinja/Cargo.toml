[package]
name = "minijinja"
version = "2.0.3"
edition = "2021"
license = "Apache-2.0"
authors = ["Armin Ronacher <armin.ronacher@active-4.com>"]
description = "a powerful template engine for Rust with minimal dependencies"
homepage = "https://github.com/mitsuhiko/minijinja"
repository = "https://github.com/mitsuhiko/minijinja"
keywords = ["jinja", "jinja2", "templates"]
readme = "README.md"
rust-version = "1.61"
exclude = ["tests"]

[package.metadata.docs.rs]
features = ["loader", "json", "urlencode", "custom_syntax", "fuel"]
rustdoc-args = ["--cfg", "docsrs", "--html-in-header", "doc-header.html"]

[features]
default = [
    "builtins",
    "debug",
    "deserialization",
    "macros",
    "multi_template",
    "adjacent_loop_items",
    "std_collections",
]

# API features
preserve_order = ["indexmap"]
deserialization = []
debug = []
loader = ["self_cell", "memo-map"]
unicode = ["unicode-ident", "unicase"]
custom_syntax = ["dep:aho-corasick"]
std_collections = []

# Speedups
key_interning = []
speedups = ["v_htmlescape"]

# Engine Features
builtins = []
macros = []
multi_template = []
adjacent_loop_items = []
fuel = []

# Extra Filters
json = ["serde_json"]
urlencode = ["percent-encoding"]

# Internal Features that should not be used
internal_debug = []
unstable_machinery = ["internal_debug"]
unstable_machinery_serde = ["unstable_machinery", "serde/derive"]

[dependencies]
aho-corasick = { version = "1.0", default-features = false, optional = true }
serde = "1.0.130"
v_htmlescape = { version = "0.15.8", optional = true }
self_cell = { version = "1.0.4", optional = true }
serde_json = { version = "1.0.68", optional = true }
percent-encoding = { version = "2.2.0", optional = true }
indexmap = { version = "1.9.0", optional = true }
memo-map = { version = "0.3.1", optional = true }
unicode-ident = { version = "1.0.5", optional = true }
unicase = { version = "2.6.0", optional = true }
stacker = { version = "0.1.15", optional = true }

[dev-dependencies]
insta = { version = "1.38.0", features = ["glob", "serde"] }
serde = { version = "1.0.130", features = ["derive"] }
serde_json = "1.0.68"
similar-asserts = "1.4.2"
