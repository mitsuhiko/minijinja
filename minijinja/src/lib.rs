//! <div align=center>
//!   <img src="https://github.com/mitsuhiko/minijinja/raw/main/artwork/logo.png" alt="" width=320>
//!   <p><strong>MiniJinja: a powerful template engine for Rust with minimal dependencies</strong></p>
//! </div>
//!
//! MiniJinja is a powerful but minimal dependency template engine for Rust which
//! is based on the syntax and behavior of the
//! [Jinja2](https://jinja.palletsprojects.com/) template engine for Python.  It's
//! implemented on top of [`serde`].  The goal is to be able to render a large
//! chunk of the Jinja2 template ecosystem from Rust with a minimal engine and to
//! leverage an already existing ecosystem of editor integrations.
//!
//! ```jinja
//! {% for user in users %}
//!   <li>{{ user.name }}</li>
//! {% endfor %}
//! ```
//!
//! You can play with MiniJinja online [in the browser
//! playground](https://mitsuhiko.github.io/minijinja-playground/) powered by a
//! WASM build of MiniJinja.
//!
//! # Why MiniJinja
//!
//! MiniJinja by its name wants to be a good default choice if you need a little
//! bit of templating with minimal dependencies.  It has the following goals:
//!
//! * Well documented, compact API
//! * Minimal dependencies, reasonable compile times and decent runtime performance
//! * Stay close as possible to Jinja2
//! * Support for expression evaluation
//! * Support for all `serde` compatible types
//! * Excellent test coverage
//! * Support for dynamic runtime objects with methods and dynamic attributes
//!
//! # Template Usage
//!
//! To use MiniJinja one needs to create an [`Environment`] and populate it with
//! templates.  Afterwards templates can be loaded and rendered.  To pass data
//! one can pass any serde serializable value.  The [`context!`] macro can be
//! used to quickly construct a template context:
//!
//! ```
//! use minijinja::{Environment, context};
//!
//! let mut env = Environment::new();
//! env.add_template("hello", "Hello {{ name }}!").unwrap();
//! let tmpl = env.get_template("hello").unwrap();
//! println!("{}", tmpl.render(context!(name => "John")).unwrap());
//! ```
//!
//! ```plain
//! Hello John!
//! ```
//!
//! For super trivial cases where you need to render a string once, you can
//! also use the [`render!`] macro which acts a bit like a replacement
//! for the [`format!`] macro.
//!
//! # Expression Usage
//!
//! MiniJinja — like Jinja2 — allows to be used as expression language.  This can be
//! useful to express logic in configuration files or similar things.  For this
//! purpose the [`Environment::compile_expression`] method can be used.  It returns
//! an expression object that can then be evaluated, returning the result:
//!
//! ```
//! use minijinja::{Environment, context};
//!
//! let env = Environment::new();
//! let expr = env.compile_expression("number < 42").unwrap();
//! let result = expr.eval(context!(number => 23)).unwrap();
//! assert_eq!(result.is_true(), true);
//! ```
//!
//! This becomes particularly powerful when [dynamic objects](crate::value::Object) are
//! exposed to templates.
//!
//! # Custom Filters
//!
//! MiniJinja lets you register functions as filter functions (see
//! [`Filter`](crate::filters::Filter)) with the engine.  These can then be
//! invoked directly from the template:
//!
//! ```
//! use minijinja::{Environment, context};
//!
//! let mut env = Environment::new();
//! env.add_filter("repeat", str::repeat);
//! env.add_template("hello", "{{ 'Na '|repeat(3) }} {{ name }}!").unwrap();
//! let tmpl = env.get_template("hello").unwrap();
//! println!("{}", tmpl.render(context!(name => "Batman")).unwrap());
//! ```
//!
//! ```plain
//! Na Na Na Batman!
//! ```
//!
//! # Learn more
//!
//! - [`Environment`]: the main API entry point.  Teaches you how to configure the environment.
//! - [`Template`]: the template object API.  Shows you how templates can be rendered.
//! - [`syntax`]: provides documentation of the template engine syntax.
//! - [`filters`]: teaches you how to write custom filters and to see the list of built-in filters.
//! - [`tests`]: teaches you how to write custom test functions and to see the list of built-in tests.
//! - [`functions`]: teaches how to write custom functions and to see the list of built-in functions.
//! - For auto reloading use the [`minijinja-autoreload`](https://docs.rs/minijinja-autoreload) crate.
//!
//! Additionally there is an [list of examples](https://github.com/mitsuhiko/minijinja/tree/main/examples)
//! with many different small example programs on GitHub to explore.
//!
//! # Error Handling
//!
//! MiniJinja tries to give you good errors out of the box.  However if you use includes or
//! template inheritance your experience will improve greatly if you ensure to render chained
//! errors.  For more information see [`Error`] with an example.
//!
//! # Size and Compile Times
//!
//! MiniJinja attempts to compile fast so it can be used as a sensible template engine choice
//! when compile times matter.  Because of this it's internally modular so unnecessary bits and
//! pieces can be removed.  The default set of features tries to strike a balance but in
//! situations where only a subset is needed (eg: `build.rs`) all default features can be
//! be disabled.
//!
//! # Optional Features
//!
//! MiniJinja comes with a lot of optional features, some of which are turned on by
//! default.  If you plan on using MiniJinja in a library, please consider turning
//! off all default features and to opt-in explicitly into the ones you actually
//! need.
//!
//! <details><summary><strong style="cursor: pointer">Configurable Features</strong></summary>
//!
//! To cut down on size of the engine some default functionality can be removed:
//!
//! - **Engine Features:**
//!
//!   - `builtins`: if this feature is removed the default filters, tests and
//!     functions are not implemented.
//!   - `macros`: when removed the `{% macro %}` tag is not included.
//!   - `multi_template`: when removed the templates related to imports and extends
//!     are removed (`{% from %}`, `{% import %}`, `{% include %}`, and `{% extends %}`
//!     as well as `{% block %}`).
//!   - `adjacent_loop_items`: when removed the `previtem` and `nextitem` attributes of
//!     the `loop` object no longer exist.  Removing this feature can provide faster
//!     template execution when a lot of loops are involved.
//!   - `unicode`: when added unicode identifiers are supported.  Without this features
//!     only ASCII identifiers can be used for variable names and attributes.
//!
//! - **Rust Functionality:**
//!
//!   - `debug`: if this feature is removed some debug functionality of the engine is
//!     removed as well.  This mainly affects the quality of error reporting.
//!   - `deserialization`: when removed this disables deserialization support for
//!     the [`Value`](crate::value::Value) type.
//!
//! There are some additional features that can be enabled:
//!
//! - `fuel`: enables the `fuel` feature which makes the engine track fuel consumption which
//!   can be used to better protect against expensive templates.
//! - `source`: enables the `Source` type which helps with dynamic loading of templates.
//! - `speedups`: enables all speedups, in particular it turns on the `v_htmlescape` dependency
//!   for faster HTML escapling.
//! - `json`: When enabled the `tojson` filter is added as builtin filter as well as
//!   the ability to auto escape via `AutoEscape::Json`.
//! - `urlencode`: When enabled the `urlencode` filter is added as builtin filter.
//! - `preserve_order`: When enable the internal value implementation uses an indexmap
//!   which preserves the original order of maps and structs.
//! - `key_interning`: if this feature is enabled the automatic string interning in
//!   the value type is enabled.  This feature used to be turned on by default but
//!   has negative performance effects in newer versions of MiniJinja since a lot of
//!   the previous uses of key interning are no longer needed.  Enabling it however
//!   cuts down on memory usage slightly in certain scenarios by interning all string
//!   keys used in dynamic map values.
//!
//! </details>
#![allow(clippy::cognitive_complexity)]
#![allow(clippy::get_first)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![deny(missing_docs)]
#![doc(html_logo_url = "https://github.com/mitsuhiko/minijinja/raw/main/artwork/logo-square.png")]

#[macro_use]
mod macros;

mod compiler;
mod defaults;
mod environment;
mod error;
mod expression;
mod key;
mod output;
mod template;
mod utils;
mod vm;

pub mod filters;
pub mod functions;
pub mod syntax;
pub mod tests;
pub mod value;

#[cfg(any(test, feature = "testutils"))]
#[cfg_attr(docsrs, doc(cfg(feature = "testutils")))]
pub mod testutils;

#[cfg(feature = "source")]
mod source;

#[cfg(feature = "debug")]
mod debug;

pub use self::defaults::{default_auto_escape_callback, escape_formatter};
pub use self::environment::Environment;
pub use self::error::{Error, ErrorKind};
pub use self::expression::Expression;
pub use self::output::Output;
pub use self::template::Template;
pub use self::utils::{AutoEscape, HtmlEscape, UndefinedBehavior};

#[cfg(feature = "source")]
pub use self::source::Source;

pub use self::macros::__context;
pub use self::vm::State;

/// This module gives access to the low level machinery.
///
/// This module is only provided by the `unstable_machinery` feature and does not
/// have a stable interface.  It mostly exists for internal testing purposes and
/// for debugging.
#[cfg(feature = "unstable_machinery")]
#[cfg_attr(docsrs, doc(cfg(feature = "unstable_machinery")))]
pub mod machinery {
    #![allow(missing_docs)]
    pub use crate::compiler::ast;
    pub use crate::compiler::codegen::CodeGenerator;
    pub use crate::compiler::instructions::{Instruction, Instructions};
    pub use crate::compiler::lexer::tokenize;
    pub use crate::compiler::parser::parse;
    pub use crate::compiler::tokens::{Span, Token};
    pub use crate::template::CompiledTemplate;
    pub use crate::vm::Vm;

    use crate::Output;

    /// Creates an [`Output`] that writes into a string.
    pub fn make_string_output(s: &mut String) -> Output<'_> {
        Output::with_string(s)
    }
}
