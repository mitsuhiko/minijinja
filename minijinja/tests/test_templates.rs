use std::collections::BTreeMap;
use std::fs;

use minijinja::{context, Environment, Error, State};

#[test]
fn test_vm() {
    let mut refs = Vec::new();
    for entry in fs::read_dir("tests/inputs/refs").unwrap() {
        let entry = entry.unwrap();
        let filename = entry.file_name();
        let filename = filename.to_str().unwrap();
        if !filename.ends_with(".txt") && !filename.ends_with(".html") {
            continue;
        }
        let source = fs::read_to_string(entry.path()).unwrap();
        refs.push((entry.path().clone(), source));
    }

    insta::glob!("inputs/*", |path| {
        if !path.metadata().unwrap().is_file() {
            return;
        }
        let filename = path.file_name().unwrap().to_str().unwrap();
        let contents = std::fs::read_to_string(path).unwrap();
        let mut iter = contents.splitn(2, "\n---\n");
        let mut env = Environment::new();
        let ctx: serde_yaml::Value = serde_yaml::from_str(iter.next().unwrap()).unwrap();

        for (path, source) in &refs {
            let ref_filename = path.file_name().unwrap().to_str().unwrap();
            env.add_template(ref_filename, source).unwrap();
        }

        env.add_template(filename, iter.next().unwrap()).unwrap();
        let template = env.get_template(filename).unwrap();

        let mut rendered = match template.render(ctx) {
            Ok(rendered) => rendered,
            Err(err) => format!("!!!ERROR!!!\n\n{:?}\n", err),
        };
        rendered.push('\n');

        insta::assert_snapshot!(&rendered);
    });
}

#[test]
fn test_custom_filter() {
    fn test_filter(_: &State, value: String) -> Result<String, Error> {
        Ok(format!("[{}]", value))
    }

    let mut ctx = BTreeMap::new();
    ctx.insert("var", 42);

    let mut env = Environment::new();
    env.add_filter("test", test_filter);
    env.add_template("test", "{{ var|test }}").unwrap();
    let tmpl = env.get_template("test").unwrap();
    let rv = tmpl.render(&ctx).unwrap();
    assert_eq!(rv, "[42]");
}

#[test]
fn test_single() {
    let mut env = Environment::new();
    env.add_template("simple", "Hello {{ name }}!").unwrap();
    let tmpl = env.get_template("simple").unwrap();
    let rv = tmpl.render(context!(name => "Peter")).unwrap();
    assert_eq!(rv, "Hello Peter!");
}
