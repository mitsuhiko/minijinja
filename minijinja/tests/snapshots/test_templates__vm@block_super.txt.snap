---
source: minijinja/tests/test_templates.rs
expression: "&rendered"
input_file: minijinja/tests/inputs/block_super.txt

---
<title>[default title]</title>
DEFAULT BODY

